test: use v2 everywhere for P2PConnection if --v2transport is enabled

mzumsande opened the pull request:
#24748 added v2 transport to the python `P2PConnection`, but so far each test that wants to make use of it needs to enable it on an individual basis.
This PR changes it so that if the test suite is run with `--v2transport` option, v2 is used in each test by default, not only for connections between two bitcoind instances as before, but also wherever `P2PConnection` is used. Individual tests can override this global option.

To do that, a few tests need to be adjusted.
In addition, I added a commit to always use v1 in a few select subtests that send a large number of large messages (e.g. large reorgs). These tests don't have a fundamental problem with v2 but become very slow due to the unoptimised python ChaCha20 implementation (~30 minutes on my computer, so probably not suitable to be run in the CI).

As a result, `python3 test_runner.py --v2transport` should succeed and use `v2` everywhere (unless v1 is chosen explicitly).

[Edit]: To make the "test each commit" CI pass, several test fixes were squashed into the last commit, which actually enables v2 p2p for `P2PConnection`. I have an unsquashed version at https://github.com/mzumsande/bitcoin/tree/202401_bip324_alltests_unsquashed, in case that helps with review.

mzumsande committed:
test: enable p2p_sendtxrcncl.py with v2transport

By adding to the test framework a wait until the v2 handshake
is completed, so that p2p_sendtxrcncl.py (which doesn't need
to be changed itself) doesnt't send out any other messages before that.

mzumsande committed:
test: enable v2 for python p2p depending on global --v2transport flag

This changes the default behavior, individual tests can overwrite this option.
As a result, it is possible to run the entire test suite with
--v2transport, and all connections to the python p2p will then use it.

Also adjust several tests that are already running with --v2transport in the
test runner (although they actually made v1 connection before this change).
This is done in the same commit so that there isn't an
intermediate commit in which the CI fails.

mzumsande committed:
test: enable p2p_invalid_messages.py with v2transport

by disabling some sub-tests that test v1-specific features,
and adapting others to v2.

mzumsande committed:
test: Don't use v2transport when it's too slow.

Sending multiple large messages is rather slow with the non-optimized python
implementation of ChaCha20.
Apart from the slowness, these tests would also run successfully with v2.

mzumsande commented:
Hmm, "test each commit fails" The problem is that commit 606f4f32014f029a6999d7f94b7231fefafdf55f (which switches `P2PConnection` to v2 by default) is enabled last. It might be best to squash everything into one big commit (even if that might make review a bit harder) or I'd have to add some temporary workarounds. Opinions?

epiccurious commented:
Testing looks good so far:
```
TEST                              | STATUS    | DURATION

feature_block.py                  | ✓ Passed  | 67 s
feature_maxuploadtarget.py        | ✓ Passed  | 58 s
p2p_ibd_stalling.py               | ✓ Passed  | 7 s
p2p_ibd_stalling.py --v2transport | ✓ Passed  | 8 s
p2p_invalid_messages.py           | ✓ Passed  | 15 s
p2p_timeouts.py                   | ✓ Passed  | 1 s
p2p_timeouts.py --v2transport     | ✓ Passed  | 1 s
p2p_v2_earlykeyresponse.py        | ✓ Passed  | 5 s
rpc_net.py                        | ✓ Passed  | 14 s
rpc_net.py --v2transport          | ✓ Passed  | 10 s

ALL                               | ✓ Passed  | 186 s (accumulated) 
Runtime: 67 s
```

(Noob question) I'm struggling to test the changes in `test_node.py`. Any Idea what I'm doing wrong here?
```
user1@comp1:~/bitcoin-pull-29358$ python3 test/functional/test_framework/test_node.py
Traceback (most recent call last):
  File "/home/user1/bitcoin-pull-29358/test/functional/test_framework/test_node.py", line 25, in <module>
    from .authproxy import (
ImportError: attempted relative import with no known parent package
user1@comp1:~/bitcoin-pull-29358$ python3 test/functional/test_runner.py test/functional/test_framework/test_node.py
Temporary test directory at /tmp/test_runner_₿_🏃_20240201_101326
WARNING! Test 'test/functional/test_framework/test_node.py' not found in full test list.
No valid test scripts specified. Check that your test is in one of the test lists in test_runner.py, or run test_runner.py with no arguments to run all tests
```

epiccurious commented:
> I added a commit disabling a few select subtests

Rather than disabling, would it make sense to keep these subtests enabled not using v2transport?

mzumsande commented:
> would reordering the commits to keep the last commit as first commit work? reviewing is definitely easier with separate commits.

That wouldn't work either, because with that last commit we actually use v2 when using P2PConnection  (if --v2transport is chosen), so all of the tests that are fixed in this PR would fail. The problem is that we already run multiple the tests in the test runner with `--v2transport` but actually use v1 with P2PConnection...
I'll try to think of other solutions. Or maybe I'll just keep it as single commits for now and when it comes closer to merging I'll squash this branch and link to a second branch here that is unsquashed but otherwise identical.

> (Noob question) I'm struggling to test the changes in test_node.py. Any Idea what I'm doing wrong here?

`test_node.py` (and the other files in the `test/functional/test_framework` subfolder) are not actual tests that can be executed directly, only the actual tests inside the `test/functional` folder can.

> Rather than disabling, would it make sense to keep these subtests enabled but just not use v2transport?

Yes, that makes sense I'll try that out!

mzumsande commented:
rebased and always use v1 instead of disabling completely in the three slow (sub)tests.

epiccurious commented:
Tested ACK.

>not actual tests that can be executed directly
Thanks for the clarification.

Certain tests take longer to run, to be expected with the added sub-tests.
```
TEST                              | STATUS    | DURATION

feature_block.py                  | ✓ Passed  | 197 s
feature_maxuploadtarget.py        | ✓ Passed  | 59 s
p2p_ibd_stalling.py               | ✓ Passed  | 59 s
p2p_ibd_stalling.py --v2transport | ✓ Passed  | 58 s
p2p_invalid_messages.py           | ✓ Passed  | 58 s
p2p_timeouts.py                   | ✓ Passed  | 1 s
p2p_timeouts.py --v2transport     | ✓ Passed  | 1 s
p2p_v2_earlykeyresponse.py        | ✓ Passed  | 4 s
rpc_net.py                        | ✓ Passed  | 22 s
rpc_net.py --v2transport          | ✓ Passed  | 19 s

ALL                               | ✓ Passed  | 478 s (accumulated) 
Runtime: 197 s
```

stratospher commented on test/functional/feature_block.py:
```diff
@@ -1263,6 +1263,10 @@ def run_test(self):
         b89a = self.update_block("89a", [tx])
         self.send_blocks([b89a], success=False, reject_reason='bad-txns-inputs-missingorspent', reconnect=True)
 
+        # Don't use v2transport for the remaining tests, the large reorg is very slow with the unoptimized python ChCha20 implementation
```
d065b33: maybe say only for `LARGE_REORG_SIZE` tests? because the last 2 tests have v2 peers after [reconnection](https://github.com/bitcoin/bitcoin/blob/d8165db8e889f423351ae35fd375c982e793d136/test/functional/feature_block.py#L1312-L1320) due to block height mismatch/invalid block header version.

stratospher reviewed (commented):
tested ACK d8165db.

mzumsande commented on test/functional/feature_maxuploadtarget.py:
```diff
@@ -67,7 +67,8 @@ def run_test(self):
 
         for _ in range(3):
-            p2p_conns.append(self.nodes[0].add_p2p_connection(TestP2PConn()))
+            # Don't use v2transport in this test (too slow with the unoptimized python ChCha20 implementation)
```
fixed

mzumsande commented:
I have addressed feedback and squashed several test fixes into the last commit, so that the "test each commit" CI is hopefully green now. I have an unsquashed version at https://github.com/mzumsande/bitcoin/tree/202401_bip324_alltests_unsquashed, in case that helps with review.

vasild reviewed (commented):
Approach ACK b9912e2df9767540ab7f53cb3ba37fd2e2491806

The code changes look good. Just the question below.

mzumsande commented on test/functional/p2p_ibd_stalling.py:
```diff
@@ -80,7 +80,8 @@ def run_test(self):
 
         # Need to wait until 1023 blocks are received - the magic total bytes number is a workaround in lack of an rpc
         # returning the number of downloaded (but not connected) blocks.
-        self.wait_until(lambda: self.total_bytes_recv_for_blocks() == 172761)
```
yes. Fixed both.

mzumsande commented:
[b9912e2 ](https://github.com/bitcoin/bitcoin/commit/b9912e2df9767540ab7f53cb3ba37fd2e2491806)to [bf5662c](https://github.com/bitcoin/bitcoin/commit/bf5662c678455fb47c402f8520215726ddfe7a94): just small fixes to the last commit message.

vasild reviewed (approved):
ACK bf5662c678455fb47c402f8520215726ddfe7a94

See also https://github.com/bitcoin/bitcoin/pull/29358#discussion_r1485626063

vasild commented on test/functional/p2p_timeouts.py:
```diff
@@ -69,11 +69,8 @@ def run_test(self):
         with self.nodes[0].assert_debug_log(['Unsupported message "ping" prior to verack from peer=0']):
             no_verack_node.send_message(msg_ping())
 
-        # With v2, non-version messages before the handshake would be interpreted as part of the key exchange.
```
Alright, "it's the same on master" that is a good point!

So, if we have `--v1transport` (for each individual test and for `test_runner.py`?) then what would happen if ran without either one of `--v1transport` or `--v2transport`?

mzumsande replied:
We'd have to make some kind of decision what to default to - maybe v1 now and switch to v2 after a majority of the network has upgraded?
My idea for now was to only add `--v1transport` in `test_runner.py` the few tests that are run twice, not everywhere.

mzumsande replied:
I have opened the draft PR #29483 with a proposal / to continue this discussion about CI and test_runner.

stratospher reviewed (commented):
reACK bf5662c6.

theStack reviewed (approved):
Tested ACK bf5662c678455fb47c402f8520215726ddfe7a94

mzumsande commented on test/functional/test_framework/test_node.py:
```diff
@@ -771,6 +773,8 @@ def addconnection_callback(address, port):
             p2p_conn.wait_for_connect()
             self.p2ps.append(p2p_conn)
 
+            if supports_v2_p2p:
+                p2p_conn.wait_until(lambda: p2p_conn.v2_state.tried_v2_handshake)
```
You are right, it's not strictly needed in `add_outbound_p2p_connection`, (though it is in `add_p2p_connection`, where the corresponding wait depends on the `send_version` option which may be unset).
I think it makes sense conceptually because it kind of decouples the v2 logic from the message logic. If we'd introduce a similar parameter `send_version`  also for `add_outbound_p2p_connection` (e.g. because we might want to craft a custom version message like `p2p_sendtxrcncl.py` does) we'd need it.

fjahr commented:
tACK bf5662c678455fb47c402f8520215726ddfe7a94
